bot/exts/internal_eval/_internal_eval.py

```diff
@@ -0,0 +1,182 @@
+            url = await self._upload_output(output)
+
+            if url:
+                upload_message = f"\nFull output here: {url}"
+            else:
+                upload_message = "\n:warning: Failed to upload full output!"
+
+            output = self.shorten_output(output)
+
+        await ctx.send(f"```py\n{output}```{upload_message}")
+
+    async def _eval(self, ctx: commands.Context, code: str) -> None:
+        """Evaluate the `code` in the current evaluation context."""
+        if code.startswith("exit"):
+            self.locals = {}
+            await ctx.send("The evaluation context was reset.")
+            return
+
+        context_vars = {
+            "message": ctx.message,
+            "author": ctx.message.author,
+            "channel": ctx.channel,
+            "guild": ctx.guild,
+            "ctx": ctx,
+            "self": self,
+            "bot": self.bot,
+            "discord": discord,
+        }
+
+        eval_context = EvalContext(context_vars, self.locals)
+
+        log.trace("Preparing the evaluation by parsing the AST of the code")
+        error = eval_context.prepare_eval(code)
+
+        if error:
+            log.trace("The code can't be evaluated due to an error")
+            await ctx.send(f"```py\n{error}\n```")
+            return
+
+        log.trace("Evaluate the AST we've generated for the evaluation")
+        new_locals = await eval_context.run_eval()
+
+        log.trace("Updating locals with those set during evaluation")
+        self.locals.update(new_locals)
+
+        log.trace("Sending the formatted output back to the context")
+        await self._send_output(ctx, eval_context.format_output())
+
+    @commands.group(name='internal', aliases=('int',))
+    @with_role(Roles.admin)
+    async def internal_group(self, ctx: commands.Context) -> None:
+        """Internal commands. Top secret!"""
+        if not ctx.invoked_subcommand:
+            await ctx.send_help(ctx.command)
+
+    @internal_group.command(name='eval', aliases=('e',))
+    @with_role(Roles.admin)
+    async def eval(self, ctx: commands.Context, *, code: str) -> None:
+        """Run eval in a REPL-like format."""
+        if match := list(FORMATTED_CODE_REGEX.finditer(code)):
+            blocks = [block for block in match if block.group("block")]
+
+            if len(blocks) > 1:
+                code = '\n'.join(block.group("code") for block in blocks)
+            else:
+                match = match[0] if len(blocks) == 0 else blocks[0]
+                code, block, lang, delim = match.group("code", "block", "lang", "delim")
+
+        else:
+            code = RAW_CODE_REGEX.fullmatch(code).group("code")
+
+        code = textwrap.dedent(code)
+        await self._eval(ctx, code)
+
+    @internal_group.command(name='reset', aliases=("clear", "exit", "r", "c"))
+    @with_role(Roles.admin)
+    async def reset(self, ctx: commands.Context) -> None:
+        """Run eval in a REPL-like format."""
```

Comment:
The docstring should be updated to give the actual purpose of the command, reset the eval context.